### PR TITLE
Add balloon_illegal test in balloon_check

### PR DESCRIPTION
**Pull request by aliang123**
qemu.tests.balloon_check:
1.remove illegal balloon value setting limitation.
2.adjust code for balloon illegal test.

qemu.tests.cfg.balloon_check:
1.add variants "balloon_illegal" for balloon illegal test.
2.add balloon_illegal test for enlarge operation

### CODE REVIEW

**Comment by aliang123** [comment]
ID:1372561

File: qemu/tests/balloon_check.py [hunk 1]
```diff
@@ -77,8 +77,15 @@ def balloon_memory(self, new_mem):
...
-        self.vm.balloon(new_mem)
+        if new_mem > self.ori_mem:
+            compare_mem = self.ori_mem
+        else:
+            compare_mem = new_mem
+        try:
+            self.vm.balloon(new_mem)
+        except:
+            if not self.params.get('illegal_value_check') and compare_mem != self.ori_mem:
+                raise Exception
```
Comment (pingl2016) [inline qemu/tests/balloon_check.py]: should give the specific error message when raise exception

**Comment by CongLi** [comment]
Hi aliang,

Please fix the pep8 problem at first.

File: qemu/tests/balloon_check.py [hunk 2]
```diff
@@ -77,8 +77,18 @@ def balloon_memory(self, new_mem):
...
+            compare_mem = self.ori_mem
+        elif new_mem == 0:
+            compare_mem = self.current_mmem
+        else:
+            compare_mem = new_mem
+        try:
+            self.vm.balloon(new_mem)
+        except Exception, e:
+            if not self.params.get('illegal_value_check') and new_mem != self.current_mmem:
+                logging.error("balloon memory fail with error message: %s" %e)
```
Comment (pingl2016) [inline qemu/tests/balloon_check.py]: better to write as:
msg = "balloon memory fail with error message: %s" %e
raise error.TestFail(msg)

File: qemu/tests/cfg/balloon_check.cfg [hunk 3]
```diff
@@ -51,23 +51,27 @@
             check_img = yes
         - balloon-stop_continue:
             sub_test_after_balloon = "stop_continue"
+        - balloon_illegal:
```
Comment (suqinhuang) [inline qemu/tests/cfg/balloon_check.cfg]: the cfg file is long, how about make a separate one?

File: qemu/tests/balloon_check.py [hunk 4]
```diff
@@ -77,8 +77,18 @@ def balloon_memory(self, new_mem):
         :type new_mem: int
         """
         error.context("Change VM memory to %s" % new_mem, logging.info)
-        compare_mem = new_mem
-        self.vm.balloon(new_mem)
+        min_sz, max_sz = self.get_memory_boundary()
```
Comment (suqinhuang) [inline qemu/tests/balloon_check.py]: min_sz and max_sz is get here, but you didn't use it

File: qemu/tests/balloon_check.py [hunk 5]
```diff
@@ -77,8 +77,18 @@ def balloon_memory(self, new_mem):
...
+        if new_mem > self.ori_mem:
```
Comment (suqinhuang) [inline qemu/tests/balloon_check.py]: do you want to compare max_sz here?

**Comment by PyLearner** [comment]
smp_16.32768m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.balloon_check.balloon_evict_and_enlarge.balloon_illegal    PASS    359

**Comment by PyLearner** [comment]
Acked-by: Meng Yang meyang@redhat.com

File: qemu/tests/balloon_check.py (hunk 4 above)
Comment (aliang123) [inline qemu/tests/balloon_check.py]: thanks for your suggestion, will correct it later.

File: qemu/tests/balloon_check.py (hunk 5 above)
Comment (aliang123) [inline qemu/tests/balloon_check.py]: No useful, will remove it.

**Comment by xiagao** [comment]
Pass on RHEL7.3 guest and Win2016 guest.
 (1/4) smp_16.2048m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.balloon_illegal: PASS (410.38 s)
 (2/4) smp_16.2048m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.balloon_illegal: PASS (333.76 s)

**Comment by xiagao** [comment]
Run sub test cases in "balloon_check" with RHEL7.3 guest and Win2016 guest, the results all PASS.

**Comment by xiagao** [comment]
Acked-by: Xiaoling Gao <xiagao@redhat.com>

File: qemu/tests/balloon_check.py [hunk 6]
```diff
@@ -77,8 +78,22 @@ def balloon_memory(self, new_mem):
         :type new_mem: int
         """
         error.context("Change VM memory to %s" % new_mem, logging.info)
-        compare_mem = new_mem
-        self.vm.balloon(new_mem)
+        try:
+            self.vm.balloon(new_mem)
+        except Exception, e:
+            if not self.params.get('illegal_value_check'):
```
Comment (CongLi) [inline qemu/tests/balloon_check.py]: true and false are all strings here.

**Comment by CongLi** [comment]
Acked-by: Cong Li <coli@redhat.com>

**Comment by pingl2016** [comment]
@aliang123 How about change your commit log as below?

qemu.tests: add negative case for balloon test

qemu.tests.balloon_check: 
1.Optimize minimum ballooned size for windows guest
2.Optimize illegal test for windows guest
3.Seperate illegal as an individual cfg file
4.Correct two pep error.
5.Correct some synax error.

qemu.tests.cfg.balloon_illegal:
1.add new cfg "balloon_illegal.cfg".
2.Adjust code for balloon_illegal test
3.Optimize code for windows's minimum balloon size

Signed-off-by: Aihua Liang <aliang@redhat.com>

**Comment by pingl2016** [comment]
Acked-by: Ping Li pingl@redhat.com

 (01/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.balloon_illegal:  PASS (395.33 s)
 (02/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.balloon_check.balloon_evict.balloon-migrate:  PASS (209.84 s)
 (03/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.balloon_check.balloon_evict.balloon-reboot:  PASS (207.27 s)
 (04/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.balloon_check.balloon_evict.balloon-shutdown:  PASS (149.40 s)
 (05/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.balloon_check.balloon_evict.balloon-stop_continue:  PASS (225.39 s)
 (06/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.balloon_check.balloon_enlarge.balloon-migrate:  PASS (207.10 s)
 (07/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.balloon_check.balloon_enlarge.balloon-reboot:  PASS (207.23 s)
 (08/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.balloon_check.balloon_enlarge.balloon-shutdown:  PASS (155.69 s)
 (09/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.balloon_check.balloon_enlarge.balloon-stop_continue:  PASS (161.60 s)
 (10/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.balloon_check.balloon_evict_and_enlarge.balloon_base:  PASS (153.52 s)
 (11/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.balloon_illegal:  PASS (355.49 s)
 (12/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.win_virtio_driver_update_test.with_balloon.before_bg_test.uninstall_install:  PASS (1051.49 s)
 (13/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.win_virtio_driver_update_test.with_balloon.before_bg_test.downgrade_upgrade:  PASS (1066.15 s)
 (14/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.win_virtio_driver_update_test.with_balloon.during_bg_test.uninstall_install:  PASS (741.22 s)
 (15/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.win_virtio_driver_update_test.with_balloon.during_bg_test.downgrade_upgrade:  PASS (758.23 s)
 (16/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.win_virtio_driver_update_test.with_balloon.after_bg_test.uninstall_install:  PASS (926.98 s)
 (17/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.win_virtio_driver_update_test.with_balloon.after_bg_test.downgrade_upgrade:  PASS (948.68 s)
 (18/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.balloon_check.balloon_evict.balloon-migrate:  PASS (395.00 s)
 (19/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.balloon_check.balloon_evict.balloon-reboot:  PASS (410.86 s)
 (20/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.balloon_check.balloon_evict.balloon-shutdown:  PASS (300.68 s)
 (21/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.balloon_check.balloon_evict.balloon-stop_continue:  PASS (306.47 s)
 (22/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.balloon_check.balloon_enlarge.balloon-migrate:  PASS (353.96 s)
 (23/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.balloon_check.balloon_enlarge.balloon-reboot:  PASS (409.12 s)
 (24/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.balloon_check.balloon_enlarge.balloon-shutdown:  PASS (302.76 s)
 (25/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.balloon_check.balloon_enlarge.balloon-stop_continue:  PASS (310.14 s)
 (26/28) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.balloon_check.balloon_evict_and_enlarge.balloon_base:  PASS (298.52 s)

 (1/2) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.balloon_hotplug: PASS (738.13 s)
 (2/2) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.balloon_hotplug: PASS (1339.45 s)

 (2/3) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.balloon_service: PASS (1604.20 s)

**Comment by aliang123** [comment]
base on patch: https://github.com/avocado-framework/avocado-vt/pull/753

File: qemu/tests/balloon_check.py [hunk 7]
```diff
@@ -332,6 +352,8 @@ def run(test, params, env):
         balloon_test = BallooningTestWin(test, params, env)
     else:
         balloon_test = BallooningTestLinux(test, params, env)
+        logging.info("Waiting about 1.5 minutes for linux guest to boot up completely")
+        time.sleep(90)
```
Comment (xutian) [inline qemu/tests/balloon_check.py]: why not vm.wait_for_login(timeout=90)

**Comment by xutian** [comment]
looks ok to me, merged, thanks @aliang123

**Pull request merged**